ddl/constraint.go

```diff
@@ -14,15 +14,267 @@
+	dbInfo, tblInfo, constraintInfo, enforced, err := checkAlterCheckConstraint(t, job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	// enforced will fetch table data and check the constraint.
+	if constraintInfo.Enforced != enforced && enforced {
+		skipCheck := false
+		failpoint.Inject("mockPassAlterConstraintCheck", func(val failpoint.Value) {
+			if val.(bool) {
+				skipCheck = true
+			}
+		})
+		if !skipCheck {
+			err = w.addTableCheckConstraint(dbInfo, tblInfo, constraintInfo, job)
+			if err != nil {
+				// check constraint error will cancel the job, job state has been changed
+				// to cancelled in addTableCheckConstraint.
+				return ver, errors.Trace(err)
+			}
+		}
+	}
+	constraintInfo.Enforced = enforced
+	ver, err = updateVersionAndTableInfoWithCheck(t, job, tblInfo, true)
+	if err != nil {
+		// update version and tableInfo error will cause retry.
+		return ver, errors.Trace(err)
+	}
+
+	job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+	return ver, nil
+}
+
+func checkDropCheckConstraint(t *meta.Meta, job *model.Job) (*model.TableInfo, *model.ConstraintInfo, error) {
+	schemaID := job.SchemaID
+	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	var constrName model.CIStr
+	err = job.DecodeArgs(&constrName)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return nil, nil, errors.Trace(err)
+	}
+
+	// do the double-check with constraint existence.
+	constraintInfo := tblInfo.FindConstraintInfoByName(constrName.L)
+	if constraintInfo == nil {
+		job.State = model.JobStateCancelled
+		return nil, nil, ErrConstraintDoesNotExist.GenWithStackByArgs(constrName)
+	}
+	return tblInfo, constraintInfo, nil
+}
+
+func checkAddCheckConstraint(t *meta.Meta, job *model.Job) (*model.DBInfo, *model.TableInfo, *model.ConstraintInfo, *model.ConstraintInfo, error) {
+	schemaID := job.SchemaID
+	dbInfo, err := t.GetDatabase(job.SchemaID)
+	if err != nil {
+		return nil, nil, nil, nil, errors.Trace(err)
+	}
+	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return nil, nil, nil, nil, errors.Trace(err)
+	}
+	constraintInfo1 := &model.ConstraintInfo{}
+	err = job.DecodeArgs(constraintInfo1)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return nil, nil, nil, nil, errors.Trace(err)
+	}
+	// do the double-check with constraint existence.
+	constraintInfo2 := tblInfo.FindConstraintInfoByName(constraintInfo1.Name.L)
+	if constraintInfo2 != nil {
+		if constraintInfo2.State == model.StatePublic {
+			// We already have a column with the same column name.
+			job.State = model.JobStateCancelled
+			return nil, nil, nil, nil, infoschema.ErrColumnExists.GenWithStackByArgs(constraintInfo1.Name)
+		}
+		// if not, that means constraint was in intermediate state.
+	}
+	return dbInfo, tblInfo, constraintInfo2, constraintInfo1, nil
```

Comment:
So can it be added again?

Reply:
Nope, 
```
dbInfo, tblInfo, constraintInfoInMeta, constraintInfoInJob, err := checkAddCheckConstraint(t, job)
```
one1 is in job from the every beginning, one2 is already in meta (mid state)
one2 is prior to one1